### PR TITLE
Fix reminders sending duplicate DELETE request and cancelling themselves

### LINKED ISSUE #795: Reminders send duplicate DELETE requests
I noticed that reminders send a duplicate DELETE request from the scheduled task. `send_reminder()` already deletes the reminder too but the task also tries to delete the reminder afterwards. `_delete_reminder()` also cancels the task after sending the request. Because a shield isn't being used, the API error ends up being suppressed (a 404 occurs because the reminder was already deleted). If not for the 404, then task being cancelled would at the least still send a warning cause the task wouldn't exist in the dictionary.

Maybe the duplicate delete exists to ensure the delete happens and task gets cancelled even if sending the reminder fails halfway. If that's indeed an extant problem which needs to be solved, then it needs to be solved differently.

Note this was discovered while working on #755, which makes the scheduler stop suppressing the errors (assuming tasks don't cancel themselves :)).

### PR DESCRIPTION
**Pull request by MarkKoz**
Fixes #795

The scheduled task will no longer try to the delete reminder. This was redundant since the `send_reminders` function, which the scheduled task calls, already deletes the reminder. The scheduled task will no longer cancel itself either; it should rely on the scheduler to take care of that.

There were race conditions between the send, delete, and edit operations. To fix them, once an operation starts on a reminder, all other operations on that reminder are blocked and aborted. This ensures multiple things cannot operate on the same resource. This was implemented through a new decorator, `mutually_exclusive`. It works by acquiring a lock for a given resource ID and checking if whether the lock is already locked. This decorator is generic enough to be re-used elsewhere. Some other clean-up and refactoring of the `decorators` module was also performed.

The decorator can optionally raise a custom exception if a resource is locked. Reminder commands are configured to raise such exceptions. The command error handler will catch them and display feedback to the user:

bild

### CODE REVIEW

File: bot/cogs/reminders.py
Comment (MarkKoz) [inline bot/cogs/reminders.py]: Because edited reminders re-use the previous ID, this delete could lead to a race condition:

1. Original reminder starts being sent
2. User edits reminder and PATCH request goes through
3. Original reminder is sent and DELETE request goes through
4. Original reminder task is cancelled (well, too late - it's already finished and been deleted) and edited one is scheduled

Now the reminder is scheduled but doesn't actually exist in the database. I wanted to put this delete into a `finally` block to ensure it's deleted even if it fails to send on Discord, but that would lead to even worse problems. In that case, the reminder would still be deleted even if the task gets cancelled.

A fix may be to first cancel the task, then send the PATCH request. However, I'm don't know how to be certain the task has been cancelled besides accessing the internal scheduler dict and polling for the task to be cancelled. The edit commands somehow need to wait for the task to be cancelled before proceeding. I'm not sure if it's even import to be so certain it's cancelled. It causing an issue may be a really small chance, though pretty sure it'd be more relevant if the delete was in the `finally` block like I mentioned earlier.

Another fix would be to create a new ID for the reminder. Then, the original reminder being deleted won't affect the edited reminder. This seems like a simple and robust solution but perhaps at the cost of convenience for users.

Ideas for solutions are welcome.

Comment (MarkKoz) [reply bot/cogs/reminders.py:143]: I've solved this by creating a decorator to abort other commands which try to use the same reminder.

File: bot/cogs/moderation/infractions.py
Comment (SebastiaanZ) [inline bot/cogs/moderation/infractions.py:233]: This call makes sense to me as I've just reviewed the previous commit. However, I'm sure that if I come back to this in a few weeks, I'm going to be puzzled by what the `2` means in the decorator factory call. Is there a way we could use a nice keyword argument here? I know you don't like long names, so something shorter than `keyword_or_positional_argument`, and maybe even the current `name_or_pos` would do.

File: bot/decorators.py
Comment (SebastiaanZ) [inline bot/decorators.py:156]: I like this. Good way to signal to the caller/error handler what's up and allow them to handle it.

File: bot/decorators.py
Comment (SebastiaanZ) [reply bot/decorators.py:151]: I created a minimal example and tried to actually get a race condition. While I still think it's possible in theory, since there's a chain of `awaits` before the sentinel `_locked` gets set to `True` in the `asyncio.Lock` instance, it does not actually seem to have the chance to create issues. We may still decide to just ditch it, since we don't need waiting logic (it's what we're trying to prevent, I think), but hmm.

Minimal example I used:
```py
from functools import wraps
import asyncio


locks = {}


def mutually_exclusive(func):
    """Minimal example of race condition."""
    @wraps(func)
    async def wrapper(example_id):
        id_ = 1
        # Get the lock for the ID. Create a lock if one doesn't exist yet.
        lock = locks.setdefault(id_, asyncio.Lock())

        if not lock.locked():
            async with lock:
                return await func(example_id)
        else:
            return 0

    return wrapper


@mutually_exclusive
async def example(example_id):
    print(f"Hello from {example_id}")
    await asyncio.sleep(1)
    print(f"Goodbye from {example_id}")
    return 1


async def main():
    tasks = 100_000
    print("Scheduling all coroutines")
    ran = sum(await asyncio.gather(*[example(i) for i in range(tasks)]))
    print(f"Done, ran {ran} tasks, stopped {tasks-ran} tasks")


asyncio.run(main())
```

File: bot/cogs/moderation/infractions.py
Comment (lemonsaurus) [reply bot/cogs/moderation/infractions.py:233]: agreed. this is very puzzling to me.

File: bot/decorators.py
Comment (SebastiaanZ) [reply bot/decorators.py:151]: A closer inspection of the source code suggests that the race condition may not occur, since the first point where the `await` chain will be suspended is after setting the sentinal `self._locked` to `True`. That said, I'm not entirely sure if my understanding of how scheduled tasks are suspended is good enough to figure this out. I think it would work like I think it does if it were `yield from` chains, which suspend at the inner `yield` subgenerator.

File: bot/decorators.py
Comment (MarkKoz) [reply bot/decorators.py:119]: Yes.

File: bot/decorators.py
Comment (MarkKoz) [reply bot/decorators.py:151]: The decorator coroutine can't yield control on synchronous code. Once one decorator starts executing, any other decorator (actually, the entire program) will be completely blocked until the first decorator reaches an await somewhere, like in `id_ = await id_` or when the lock is acquired. If there was an await between the if condition and the `async with`, then that could indeed cause a race condition. But this is not the case here since it immediately tries to enter the context manager.

Comment (MarkKoz) [reply bot/decorators.py:151]: But you're right, a sentinel is more appropriate. I just didn't feel like creating a custom object when the lock can do the job too. Do you think it's worth using your example to be clearer?

File: bot/decorators.py
Comment (MarkKoz) [reply bot/decorators.py]: Oops, I don't know what I was thinking.

File: bot/decorators.py
Comment (MarkKoz) [reply bot/decorators.py]: It made more sense in previous iterations of the code. I agree with your concern, but on the other hand, it is nice to have a variable for it that keeps it consistent for all uses. I don't know. It's short and simple enough of a type that I'm on the fence.

File: bot/decorators.py
Comment (SebastiaanZ) [reply bot/decorators.py:151]: > The decorator coroutine can't yield control on synchronous code. Once one decorator starts executing, any other decorator (actually, the entire program) will be completely blocked until the first decorator reaches an await somewhere, like in `id_ = await id_` or when the lock is acquired. If there was an await between the if condition and the `async with`, then that could indeed cause a race condition. But this is not the case here since it immediately tries to enter the context manager.

The line that acquires the lock:
```py
async with lock:
```
*can* yield back control to event loop, as it has multiple awaits within it, like `await self.acquire()`. This means that if every await has the potential to yield back control, control may be yielded back before the `self._locked = True` is run.

Now, I think that because each coroutine is primed when it's scheduled (it runs to the first await), we're still safe, but I'm not entirely sure if the `yield from` comparison/analogy holds.

File: bot/decorators.py
Comment (MarkKoz) [reply bot/decorators.py]: Ah, it's cause PyCharm incorrectly determined a return annotation for `check()` and I was relying on that.

File: bot/decorators.py
Comment (SebastiaanZ) [reply bot/decorators.py:151]: To make the implicit `await` in the context manager a bit more explicit, this is the equivalence code, according the docs:

```py
await lock.acquire()  # <- this is where the await is hiding
try:
    return await func(example_id)
finally:
    lock.release()
```

With the context manager, there are two coroutines involved in the context manager: `__aenter__`, which then awaits `Lock.acquire`.

Comment (MarkKoz) [reply bot/decorators.py:151]: That's correct, but this works based on the assumption that the lock will be immediately acquired if it's available, which is correct:

```py
    async def acquire(self):
        if (not self._locked and (self._waiters is None or
                all(w.cancelled() for w in self._waiters))):
            self._locked = True
            return True
```

I get your question now. I'm not sure, but I don't think that merely an `await` can yield control. I think the coroutine has to actually be doing something asynchronous, like using a future.

Comment (SebastiaanZ) [reply bot/decorators.py:151]: > I get your question now. I'm not sure, but I don't think that merely an await can yield control. I think the coroutine has to actually be doing something asynchronous, like using a future.

That's the main question I had, yes. If my generator-based analogy with `yield from` holds, scheduling a task would mean running it to the inner most `await` that is waiting for a future. If that's the case, our acquisition of the lock would happen synchronously, as we hit upon this `if` before any such suspension of the task happens. I'm just not entirely sure that is how it works.

**Comment by aeros** [comment]
Sorry for the delay, I've been a bit otherwise occupied recently. I'll try to set aside some time to provide a review for this in the next few days.

File: bot/decorators.py
Comment (MarkKoz) [reply bot/decorators.py]: I've gone ahead with your suggestion to use a custom lock class.

File: bot/decorators.py
Comment (SebastiaanZ) [reply bot/decorators.py]: I don't mind it that much. I think it's easy enough to look this up to retain the documentation power of our function annotations.

**Review by SebastiaanZ — APPROVED** [review]
Long overdue review, but the code looks solid. It does add some more complicated features to our bot, but I think the "this should only ever happen once so if something else tries it, it should just stop" behavior is something that come in handy with different features as well.

I tested the functionality, it all seems to work, although that obviously doesn't tell us everything, which a fairly complicated feature like this. I think the utilities introduced are a candidate to receive unit tests in the future, although the tests will probably be slightly difficult to write.

File: bot/decorators.py
Comment (kwzrd) [inline bot/decorators.py:126]: I had to look around to see what `function` is. It gives the impression of being a parameter or a local variable, not an imported module. I'm not sure what a better name would be though.

**Review by kwzrd — APPROVED** [review]
I'm so sorry it took this long to get reviews. I tried my best, but I'm not able to approve as confidently as I'd like.

Conceptually, the approach taken makes sense to me. The functionality is exposed via a very nice API, but I found the internals to be fairly complex in some places, and I had trouble "running" the code in my head. My testing indicates that everything is functional and that this indeed resolves #795.

I will leave it to you whether you trust my review or would prefer to wait for someone more qualified to look over the code.

By the way, if I wanted to apply `lock_arg` twice (to lock a function based on two distinct args), that would be possible, right?

File: bot/decorators.py
Comment (MarkKoz) [reply bot/decorators.py:126]: It could be qualified with `utils.` if you find that more readable.

File: bot/utils/lock.py
Comment (MarkKoz) [reply bot/utils/lock.py:31]: Probably both. It could be a property. I think I just based this off the code Ves wrote in a comment.

File: bot/utils/lock.py
Comment (MarkKoz) [reply bot/utils/lock.py:52]: I wanted this to be very flexible. There's no technical reason any hashable type shouldn't be allowed, and I think having prettier logs isn't an excuse for such restriction.

Comment (MarkKoz) [reply bot/utils/lock.py:52]: Well, this actually does get shown to users via the `LockedResourceError`. So maybe you have a point.

File: bot/utils/lock.py
Comment (MarkKoz) [reply bot/utils/lock.py:93]: > With locks being a WeakValueDictionary, after lock goes out of scope here, the instance will be removed from the dict and garbage collected, correct?

Not quite. It will get garbage collected once there are no more decorators holding references to the lock instance. The dictionary itself doesn't count as something holding a reference.

> The local lock shadows the outer function name. I think that the function could have a more descriptive name. It's only intended to be used from within this module, right?

No, the `lock` function is a perfectly valid decorator on its own. It's useful if you need to lock on a constant value rather than a dynamic one via an argument.

I'll find a non-shadowing name for the variable.

File: bot/decorators.py
Comment (MarkKoz) [reply bot/decorators.py:126]: I have to import it in `utils/__init__.py` to support that. I tried it out and it felt too verbose, just cause of the length.

Comment (MarkKoz) [reply bot/decorators.py:126]: `import function as func_utils` is another option, but feels redundant to do it everywhere the module is imported. On the other hand, it'd also feel redundant to actually name the module `func_utils` when it's already in the `utils` package.

File: bot/utils/lock.py
Comment (kwzrd) [reply bot/utils/lock.py]: > Not quite. It will get garbage collected once there are no more decorators holding references to the lock instance.

Ok, that's what I was thinking but failed to express properly.

File: bot/decorators.py
Comment (kwzrd) [reply bot/decorators.py:131]: These are exactly the options I considered but didn't like any of them enough to propose them. Eh, it's fine the way it is, it's just the name `function` within the context of a decorator misleads, but it's not like you can't just look up where the name is defined.

File: bot/utils/lock.py
Comment (kwzrd) [reply bot/utils/lock.py]: > No, the `lock` function is a perfectly valid decorator on its own. It's useful if you need to lock on a constant value rather than a dynamic one via an argument.

I did not realize that, thanks for explaining.

**Pull request merged**
